steward/src/cellars.rs

```diff
@@ -61,8 +90,86 @@ pub fn normalize_address(address: String) -> String {
         .to_string();
 }
 
+// validation logic
+
+pub fn check_blocked_adaptor(adaptor_id: &str) -> Result<(), Error> {
+    let adaptor_id = normalize_address(adaptor_id.to_string());
+    if BLOCKED_ADAPTORS.contains(&adaptor_id.as_str()) {
+        return Err(sp_call_error(format!("adaptor {adaptor_id} is blocked")));
+    }
+
+    Ok(())
+}
+
+pub fn check_blocked_position(position: u32) -> Result<(), Error> {
+    if BLOCKED_POSITIONS.contains(&position) {
+        return Err(sp_call_error(format!("position {position} is blocked")));
+    }
+
+    Ok(())
+}
+
+pub fn validate_add_position(cellar_id: &str, position: u32) -> Result<(), Error> {
+    check_blocked_position(position)?;
+    let cellar_id = normalize_address(cellar_id.to_string());
+    if !ALLOWED_ADD_POSITION.contains(&(&cellar_id, position)) {
```

Comment:
a). `_::contains()` takes a `&T`. For `BLOCKED_POSITIONS`, `T` is `u32`, and for `ALLOWED_ADD_POSITION`, `T` is a tuple `(&_, _)`. 

In both collections the `position` value is owned by the collection.

b) This is a qwerk of the same `contains()` signature. Since `as_str()` spits out a `&str`, the contains call wants a `&&str`. The same thing is happening on this line, but the deref of the `String` is inside of the tuple.

For some reason, when a `&str` is required, the compiler is ok with using `&String`, but when a `&&str` is wanted, it will not accept `&&String` or `&String`, thus the `.as_str()` call.